### ISSUE TITLE
Support for filtering by annotation

### ISSUE THREAD

**Issue by PhilippWendler** [spotbugs/spotbugs#543]
Please add support for filtering by annotations to the SpotBugs filter file. Such a filter could for example look like
```
<Match>
  <AnnotatedWith name="annotation.class.name"/>
</Match>
```
Of course it should be possible to combine this filter with other existing filters. The annotation filter should work on packages, classes, methods, fields, and local variables, basically every Java element for which already a filter exists.

More specific filters (like annotated with a specific annotation and a specific value of this annotation) would be possible, but also more complicated, thus I suggest to leave them out for now.

Of course this annotation filter would match only annotations that have a retention level of at least `class`, but there are still enough use cases for such a filter. For example, we have annotations that mark elements that are accessed via reflection, and because SpotBugs does not know about this reflective accesses there are false alarms for some checks. Now we have to disable these checks globally, with such a filter we could disable the checks only where they need to be disabled.

For completeness: related to #464 and [feature request 256 of FindBugs](https://sourceforge.net/p/findbugs/feature-requests/256/), but I was asked to create a new issue.

**Comment by oujesky** [comment]
Just ran to an use case for this - JavaFX views supports private event handler methods to be annotated with `@FXML` annotation, SpotBugs hovewer marks such methods with a `UPM_UNCALLED_PRIVATE_METHOD` rule violation.

**Comment by mannodermaus** [comment]
I've recently been thinking about what the syntax for annotation support in Spotbugs might look like, and I'd like to get a discussion going with the current findings.

Assume the following annotation:

```java
@Target(ElementType.TYPE)
@Retention(RetentionPolicy.CLASS)
public @interface MyAnnotation {
    String value();
    int counter() default 0;
}
```

Applied on these classes:

```java
@MyAnnotation("Test")
public void ExampleFragment1 {
}

@MyAnnotation(value = "Test", counter = 42)
public void ExampleFragment2 {
}
```

The two approaches I was able to come up with are detailed in the filter file description below. Each scenario is supposed to match the class named inside the `<Class>` element (please disregard the matched `pattern`, that's just for demonstration).

```xml
<?xml version="1.0" encoding="UTF-8"?>
<FindBugsFilter>
    <Match>
        <Class name="~.*\.*Fragment" />
        <Bug pattern="UWF_FIELD_NOT_INITIALIZED_IN_CONSTRUCTOR" />
    </Match>

    <!-- 
        Approach 1: New attributes in Class, Method, Field, Local
    -->

    <!-- Scenario 1: Matching class with annotation, using FQCN -->
    <Match>
        <Class name="~.*\.*Fragment1" annotationName="com.example.MyAnnotation" />
        <Bug pattern="UWF_FIELD_NOT_INITIALIZED_IN_CONSTRUCTOR" />
    </Match>

    <!-- Scenario 2: Matching class with annotation, using regular expression -->
    <Match>
        <Class name="~.*\.*Fragment1" annotationName="~.*\.MyAnnotation" />
        <Bug pattern="UWF_FIELD_NOT_INITIALIZED_IN_CONSTRUCTOR" />
    </Match>

    <!-- Scenario 3: Matching class with annotation having specific value (implied "value()" element) -->
    <Match>
        <Class name="~.*\.*Fragment1" annotationName="~.*\.MyAnnotation" annotationValue="Test" />
        <Bug pattern="UWF_FIELD_NOT_INITIALIZED_IN_CONSTRUCTOR" />
    </Match>

    <!-- Scenario 4: Matching class with annotation having specific value ("value()" element) -->
    <Match>
        <Class name="~.*\.*Fragment1" annotationName="~.*\.MyAnnotation" annotationElement="value" annotationValue="Test" />
        <Bug pattern="UWF_FIELD_NOT_INITIALIZED_IN_CONSTRUCTOR" />
    </Match>

    <!-- Scenario 5: Matching class with annotation having specific value ("counter()" element with default value) -->
    <Match>
        <Class name="~.*\.*Fragment1" annotationName="~.*\.MyAnnotation" annotationElement="counter" annotationValue="0" />
        <Bug pattern="UWF_FIELD_NOT_INITIALIZED_IN_CONSTRUCTOR" />
    </Match>

    <!-- Scenario 6: Matching class with annotation having specific value ("counter()" element with explicit value) -->
    <Match>
        <Class name="~.*\.*Fragment2" annotationName="~.*\.MyAnnotation" annotationElement="counter" annotationValue="42" />
        <Bug pattern="UWF_FIELD_NOT_INITIALIZED_IN_CONSTRUCTOR" />
    </Match>

    <!-- Scenario 7: Matching class with annotation having multiple values -->
    <Match>
        <Class name="~.*\.*Fragment2" annotationName="~.*\.MyAnnotation2" annotationValue="Test" />
        <Class name="~.*\.*Fragment2" annotationName="~.*\.MyAnnotation2" annotationElement="counter" annotationValue="42" />
        <Bug pattern="UWF_FIELD_NOT_INITIALIZED_IN_CONSTRUCTOR" />
    </Match>

    <!-- 
        Approach 2: Child elements for Class, Method, Field, Local
    -->

    <!-- Scenario 1: Matching class with annotation, using FQCN -->
    <Match>
        <Class name="~.*\.*Fragment1">
            <Annotation name="com.example.MyAnnotation" />
        </Class>
        <Bug pattern="UWF_FIELD_NOT_INITIALIZED_IN_CONSTRUCTOR" />
    </Match>

    <!-- Scenario 2: Matching class with annotation, using regular expression -->
    <Match>
        <Class name="~.*\.*Fragment1">
            <Annotation name="~.*\.MyAnnotation" />
        </Class>
        <Bug pattern="UWF_FIELD_NOT_INITIALIZED_IN_CONSTRUCTOR" />
    </Match>

    <!-- Scenario 3: Matching class with annotation having specific value (implied "value()" element) -->
    <Match>
        <Class name="~.*\.*Fragment1">
            <Annotation name="~.*\.MyAnnotation" value="Test" />
        </Class>
        <Bug pattern="UWF_FIELD_NOT_INITIALIZED_IN_CONSTRUCTOR" />
    </Match>

    <!-- Scenario 4: Matching class with annotation having specific value ("value()" element) -->
    <Match>
        <Class name="~.*\.*Fragment1">
            <Annotation name="~.*\.MyAnnotation" element="value" value="Test" />
        </Class>
        <Bug pattern="UWF_FIELD_NOT_INITIALIZED_IN_CONSTRUCTOR" />
    </Match>

    <!-- Scenario 5: Matching class with annotation having specific value ("counter()" element with default value) -->
    <Match>
        <Class name="~.*\.*Fragment1">
            <Annotation name="~.*\.MyAnnotation" element="counter" value="0" />
        </Class>
        <Bug pattern="UWF_FIELD_NOT_INITIALIZED_IN_CONSTRUCTOR" />
    </Match>

    <!-- Scenario 6: Matching class with annotation having specific value ("counter()" element with explicit value) -->
    <Match>
        <Class name="~.*\.*Fragment2">
            <Annotation name="~.*\.MyAnnotation" element="counter" value="42" />
        </Class>
        <Bug pattern="UWF_FIELD_NOT_INITIALIZED_IN_CONSTRUCTOR" />
    </Match>

    <!-- Scenario 7: Matching class with annotation having multiple values -->
    <Match>
        <Class name="~.*\.*Fragment2">
            <Annotation name="~.*\.MyAnnotation" value="Test" />
            <Annotation name="~.*\.MyAnnotation" element="counter" value="42" />
        </Class>
        <Bug pattern="UWF_FIELD_NOT_INITIALIZED_IN_CONSTRUCTOR" />
    </Match>
</FindBugsFilter>
```

Personally, I'm leaning more towards Approach 2. What could be problems with these concepts - anything we're missing here? Alternatives to writing this? I'd love to hear some feedback.

**Comment by gtoison** [comment]
So this was fixed by #2395 and released in 4.8.0
Please note that the annotation needs to be retained in the .class file, because SpotBugs analyzes the compiled bytecode, not the source files.
